File: BEmuJava/bemu/src/main/java/com/bemu/BEmu/MarketDataRequest/EventMarket.java
//------------------------------------------------------------------------------
// <copyright project="BEmuJava" file="BEmu.MarketDataRequest.EventMarket.java" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

package com.bemu.BEmu.MarketDataRequest;

import java.util.Map;
import java.util.List;
import java.util.ArrayList;

import com.bemu.BEmu.Event;
import com.bemu.BEmu.Message;
import com.bemu.BEmu.MessageIterator;
import com.bemu.BEmu.Subscription;
import com.bemu.BEmu.CorrelationID;

public class EventMarket extends Event
{
	private final List<Message> _messages;
	
	//don't use this.  It's not in the Bloomberg API
	public EventMarket(EventType evtType, CorrelationID corr, List<Subscription> subscriptions) throws Exception //use for subscribing
    {
        this._messages = new ArrayList<Message>();        
        
        switch (evtType.intValue())
        {        
            case Event.EventType.Constants.SESSION_STATUS:
            	super._eventType = evtType;
                MessageMarketSessionOpened msgSessionOpened = new MessageMarketSessionOpened();
                this._messages.add(msgSessionOpened);
                break;

            case Event.EventType.Constants.SERVICE_STATUS:
                super._eventType = evtType;
                MessageMarketServiceStatus msgServiceStatus = new MessageMarketServiceStatus(corr);
                this._messages.add(msgServiceStatus);
                break;

            case Event.EventType.Constants.SUBSCRIPTION_STATUS:
                super._eventType = evtType;
                
                for(int i = 0; i < subscriptions.size(); i++)
                {
                	Subscription item = subscriptions.get(i);
                	
                    MessageMarketSubscriptionStarted msgSubStatus = new MessageMarketSubscriptionStarted(item);
                    this._messages.add(msgSubStatus); 
                }
                break;

            case Event.EventType.Constants.SUBSCRIPTION_DATA:
                super._eventType = evtType;
                
                for(int i = 0; i < subscriptions.size(); i++)
                {
                	Subscription item = subscriptions.get(i);
                	
                    MessageMarketSubscriptionData msgSubData = new MessageMarketSubscriptionData(item, EventMarket.GenerateFakeMessageData(item));
                    this._messages.add(msgSubData);
                }
                break;

            default:
                throw new Exception(String.format("BEmu.MarketDataRequest.EventMarket.EventMarket: doesn't expect EventType %s", evtType.toString()));
        }
    }
	
    private static Map<String, Object> GenerateFakeMessageData(Subscription sub)
    {
    	return com.bemu.BEmu.types.RandomDataGenerator.getMarketDataFields(sub.fields());
    }
    
	public MessageIterator messageIterator()
	{
		return new MessageIterator(this._messages);
	}
    
}

File: BEmu/MarketDataRequest/EventMarket.cs
//------------------------------------------------------------------------------
// <copyright project="BEmu" file="MarketDataRequest/EventMarket.cs" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

namespace BEmu.MarketDataRequest
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Text;

    internal class EventMarket : Event
    {
        private readonly List<Message> _messages;

        internal EventMarket(EventType evtType, CorrelationID corr, IEnumerable<Subscription> subscriptions) //use for subscribing
        {
            this._messages = new List<Message>();

            switch (evtType)
            {
                case EventType.SESSION_STATUS:
                    base._type = evtType;
                    MessageMarketSessionOpened msgSessionOpened = new MessageMarketSessionOpened();
                    this._messages.Add(msgSessionOpened);
                    break;

                case EventType.SERVICE_STATUS:
                    base._type = evtType;
                    MessageMarketServiceStatus msgServiceStatus = new MessageMarketServiceStatus(corr);
                    this._messages.Add(msgServiceStatus);
                    break;

                case EventType.SUBSCRIPTION_STATUS:
                    base._type = evtType;
                    foreach (var item in subscriptions)
                    {
                        bool securityError = Types.Rules.IsSecurityError(item.Security);
                        if (securityError)
                        {
                            MessageMarketSubscriptionFailure msgError = new MessageMarketSubscriptionFailure(item);
                            this._messages.Add(msgError);
                        }
                        else
                        {
                            MessageMarketSubscriptionStarted msgSubStatus = new MessageMarketSubscriptionStarted(item);
                            this._messages.Add(msgSubStatus);
                        }
                    }
                    break;

                case EventType.SUBSCRIPTION_DATA:
                    base._type = evtType;
                    foreach (var item in subscriptions)
                    {
                        bool securityError = Types.Rules.IsSecurityError(item.Security);
                        if (!securityError)
                        {
                            MessageMarketSubscriptionData msgSubData = new MessageMarketSubscriptionData(item, EventMarket.GenerateFakeMessageData(item));
                            this._messages.Add(msgSubData);
                        }
                    }
                    break;

                default:
                    throw new NotImplementedException(string.Format("BEmu.MarketDataRequest.EventMarket.EventMarket: doesn't expect EventType {0}", evtType.ToString()));
            }
        }

        internal EventMarket(EventType evtType, Subscription subscription) //use for unsubscribing
        {
            this._messages = new List<Message>();

            switch (evtType)
            {
                case EventType.SUBSCRIPTION_STATUS:
                    base._type = evtType;
                    MessageMarketSubscriptionCanceled msgCancel = new MessageMarketSubscriptionCanceled(subscription);
                    this._messages.Add(msgCancel);
                    break;
                default:
                    break;
            }
        }

        private static Dictionary<string, object> GenerateFakeMessageData(Subscription sub)
        {
            return Types.RandomDataGenerator.GetMarketDataFields(sub.Fields);
        }

        public override IEnumerable<Message> GetMessages()
        {
            return this._messages;
        }
    }
}
